File: src/upstream/providers/opencode-go.ts
import { createOpenAICompatibleProvider } from "./openai-compatible";
import type { ProviderModelEntry } from "./models";

// Curated subset of the live Go catalog, the full, current list is always
// available via the console's "Fetch models" action. No pricing: this is
// OpenCode's own bundled subscription catalog, not raw per-token vendor
// pricing (each id is OpenCode's own re-branding, not a stable models.dev key).
export const opencodeGoModels: ProviderModelEntry[] = [
  { id: "grok-4.5", reasoning: true, contextWindow: 131072, maxOutputTokens: 16384, pricing: { input: 2, output: 6 } },
  { id: "glm-5.2", reasoning: true, contextWindow: 1000000, maxOutputTokens: 16384, pricing: { input: 1.4, output: 4.4 } },
  { id: "kimi-k3", reasoning: true, vision: true, contextWindow: 1000000, maxOutputTokens: 128000, pricing: { input: 3, output: 15 } },
  { id: "kimi-k2.7-code", reasoning: true, contextWindow: 262144, maxOutputTokens: 64000, pricing: { input: 0.95, output: 4 } },
  { id: "mimo-v2.5-pro", reasoning: true, contextWindow: 1048576, maxOutputTokens: 131072, pricing: { input: 0.435, output: 0.87 } },
  { id: "qwen3.7-max", reasoning: true, contextWindow: 1000000, maxOutputTokens: 16384, pricing: { input: 2.5, output: 7.5 } },
  { id: "minimax-m3", reasoning: true, contextWindow: 256000, maxOutputTokens: 32768, pricing: { input: 0.3, output: 1.2 } },
  { id: "deepseek-v4-pro", reasoning: true, contextWindow: 1000000, maxOutputTokens: 384000, pricing: { input: 0.435, output: 0.87 } },
  { id: "deepseek-v4-flash", reasoning: true, contextWindow: 1000000, maxOutputTokens: 384000, pricing: { input: 0.14, output: 0.28 } },
  { id: "hy3", reasoning: true, contextWindow: 262144, maxOutputTokens: 16384, pricing: { input: 0.14, output: 0.58 } },
];

export const opencodeGoProvider = createOpenAICompatibleProvider({
  id: "opencode-go",
  name: "OpenCode Go",
  icon: "opencode-go",
  baseUrl: "https://opencode.ai/zen/go/v1",
  credentialUrl: "https://opencode.ai/auth",
  models: opencodeGoModels,
});

File: router/internal/providers/builtin/opencodego.go
package builtin

func OpenCodeGo() ProviderDefinition {
	return ProviderDefinition{ID: "opencodego", DisplayName: "OpenCode Go", Protocol: ProtocolOpenAI, Adapter: AdapterOpenAI, CredentialKind: CredentialAPIKey, BaseURL: "https://opencode.ai/zen/go/v1", Surfaces: []Surface{SurfaceOpenAIChat}, ModelsDevID: "opencode-go", Models: []ProviderModel{
		{ID: "deepseek-v4-flash", DisplayName: "DeepSeek V4 Flash", ContextWindow: 1000000, MaxOutput: 384000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "deepseek-v4-pro", DisplayName: "DeepSeek V4 Pro", ContextWindow: 1000000, MaxOutput: 384000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "glm-5", DisplayName: "GLM-5", ContextWindow: 202752, MaxOutput: 32768, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "glm-5.1", DisplayName: "GLM-5.1", ContextWindow: 202752, MaxOutput: 32768, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "glm-5.2", DisplayName: "GLM-5.2", ContextWindow: 1000000, MaxOutput: 131072, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "grok-4.5", DisplayName: "Grok 4.5", ContextWindow: 500000, MaxOutput: 500000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "hy3", DisplayName: "Hy3", ContextWindow: 256000, MaxOutput: 64000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "kimi-k2.5", DisplayName: "Kimi K2.5", ContextWindow: 262144, MaxOutput: 65536, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "kimi-k2.6", DisplayName: "Kimi K2.6", ContextWindow: 262144, MaxOutput: 65536, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "kimi-k2.7-code", DisplayName: "Kimi K2.7 Code", ContextWindow: 262144, MaxOutput: 262144, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "kimi-k3", DisplayName: "Kimi K3 (2x usage)", ContextWindow: 1048576, MaxOutput: 131072, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "mimo-v2-omni", DisplayName: "MiMo-V2-Omni", ContextWindow: 262144, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "mimo-v2-pro", DisplayName: "MiMo-V2-Pro", ContextWindow: 1048576, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "mimo-v2.5", DisplayName: "MiMo V2.5", ContextWindow: 1000000, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "mimo-v2.5-pro", DisplayName: "MiMo V2.5 Pro", ContextWindow: 1048576, MaxOutput: 128000, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "minimax-m2.5", DisplayName: "MiniMax-M2.5", ContextWindow: 204800, MaxOutput: 65536, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "minimax-m2.7", DisplayName: "MiniMax-M2.7", ContextWindow: 204800, MaxOutput: 131072, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "minimax-m3", DisplayName: "MiniMax-M3", ContextWindow: 1000000, MaxOutput: 131072, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "qwen3.5-plus", DisplayName: "Qwen3.5 Plus", ContextWindow: 262144, MaxOutput: 65536, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "qwen3.6-plus", DisplayName: "Qwen3.6 Plus", ContextWindow: 1000000, MaxOutput: 65536, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
		{ID: "qwen3.7-max", DisplayName: "Qwen3.7 Max", ContextWindow: 1000000, MaxOutput: 65536, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: false}},
		{ID: "qwen3.7-plus", DisplayName: "Qwen3.7 Plus", ContextWindow: 1000000, MaxOutput: 65536, Capabilities: &ProviderCaps{Surfaces: []Surface{SurfaceOpenAIChat}, Streaming: true, Reasoning: true, ToolCalls: false, Images: true}},
	}}
}
